app/src/main/java/org/wikipedia/diff/ArticleEditDetailsViewModel.kt

```diff
@@ -15,67 +14,128 @@ import org.wikipedia.dataclient.mwapi.MwQueryResponse
             revisionDetails.postValue(Resource.Error(throwable))
         }) {
             withContext(Dispatchers.IO) {
-                val response = ServiceFactory.get(title.wikiSite).getRevisionDetailsAscending(title.prefixedText, 2, revisionId)
-                val revisions = response.query?.firstPage()!!.revisions
-                if (revisions.isNotEmpty()) {
-                    revisionDetails.postValue(Resource.Success(revisions))
+                if (revisionIdFrom >= 0) {
+                    var responseFrom: MwQueryResponse? = null
+                    var responseTo: MwQueryResponse? = null
+                    coroutineScope {
+                        launch {
+                            responseFrom = ServiceFactory.get(pageTitle.wikiSite).getRevisionDetailsWithInfo(pageTitle.prefixedText, 2, revisionIdFrom)
+                        }
+                        launch {
+                            responseTo = ServiceFactory.get(pageTitle.wikiSite).getRevisionDetailsWithInfo(pageTitle.prefixedText, 2, revisionIdTo)
+                        }
+                    }
+
+                    val pageTo = responseTo?.query?.firstPage()!!
+                    revisionFrom = responseFrom?.query?.firstPage()!!.revisions[0]
+                    revisionTo = pageTo.revisions[0]
+                    canGoForward = revisionTo!!.revId < pageTo.lastrevid
+                } else {
+                    val response = ServiceFactory.get(pageTitle.wikiSite).getRevisionDetailsWithInfo(pageTitle.prefixedText, 2, revisionIdTo)
+                    val page = response.query?.firstPage()!!
+                    val revisions = page.revisions
+                    revisionTo = revisions[0]
+                    canGoForward = revisions[0].revId < page.lastrevid
+                    revisionFrom = if (revisions.size > 1) { revisions[1] } else null
                 }
+
+                revisionToId = revisionTo!!.revId
+                revisionFromId = if (revisionFrom != null) revisionFrom!!.revId else revisionTo!!.parentRevId
+
+                revisionDetails.postValue(Resource.Success(Unit))
+                getDiffText(revisionFromId, revisionToId)
+            }
+        }
+    }
+
+    fun goBackward() {
+        revisionToId = revisionFromId
+        getRevisionDetails(revisionToId)
+    }
+
+    fun goForward() {
+        val revisionIdFrom = revisionToId
+        viewModelScope.launch(CoroutineExceptionHandler { _, throwable ->
+            revisionDetails.postValue(Resource.Error(throwable))
+        }) {
+            withContext(Dispatchers.IO) {
+                val response = ServiceFactory.get(pageTitle.wikiSite).getRevisionDetailsAscending(pageTitle.prefixedText, 2, revisionIdFrom)
+                val page = response.query?.firstPage()!!
+                val revisions = page.revisions
+
+                revisionFrom = revisions[0]
+                revisionTo = if (revisions.size > 1) { revisions[1] } else revisions[0]
```

Comment:
We can do this `revisions.getOrElse(1) { revisions.first() }`

Reply:
yep!